testing/async/test_policy_overloads.h

```diff
@@ -0,0 +1,421 @@
+  }
+
+  static void test_postfix_overloads(std::size_t const,
+                                     size_const<num_postfix_arg_sets>)
+  {
+    // terminal case, no-op
+  }
+
+  //----------------------------------------------------------------------------
+  // For the specified postfix overload set, test the algorithm with several
+  // different policy configurations.
+  template <std::size_t postfix_idx>
+  static void run_basic_policy_tests(std::size_t const num_values)
+  {
+    // When a policy uses the default stream, the algorithm implementation
+    // should spawn a new stream in the returned event:
+    auto using_default_stream = [](auto& e) {
+      ASSERT_NOT_EQUAL(thrust::cuda_cub::default_stream(),
+                       e.stream().native_handle());
+    };
+
+    // When a policy uses a non-default stream, the implementation should pass
+    // the stream through to the output:
+    thrust::system::cuda::detail::unique_stream test_stream{};
+    auto using_test_stream = [&test_stream](auto& e) {
+      ASSERT_EQUAL(test_stream.native_handle(), e.stream().native_handle());
+    };
+
+    // Test the different types of policies:
+    run_basic_policy_test<postfix_idx>("(no policy)",
+                                       std::make_tuple(),
+                                       using_default_stream,
+                                       num_values);
+
+    run_basic_policy_test<postfix_idx>("thrust::device",
+                                       std::make_tuple(thrust::device),
+                                       using_default_stream,
+                                       num_values);
+
+    run_basic_policy_test<postfix_idx>(
+      "thrust::device(thrust::device_allocator<void>{})",
+      std::make_tuple(thrust::device(thrust::device_allocator<void>{})),
+      using_default_stream,
+      num_values);
+
+    run_basic_policy_test<postfix_idx>("thrust::device.on(test_stream.get())",
+                                       std::make_tuple(
+                                         thrust::device.on(test_stream.get())),
+                                       using_test_stream,
+                                       num_values);
+
+    run_basic_policy_test<postfix_idx>(
+      "thrust::device(thrust::device_allocator<void>{}).on(test_stream.get())",
+      std::make_tuple(
+        thrust::device(thrust::device_allocator<void>{}).on(test_stream.get())),
+      using_test_stream,
+      num_values);
+  }
+
+  // Wrap the basic_policy_test in a try/catch block to inject helpful info into
+  // any thrown unit test exceptions.
+  template <std::size_t postfix_idx, typename PolicyTuple, typename ValidateEvent>
+  static void run_basic_policy_test(std::string const &policy_desc,
+                                    PolicyTuple &&policy_tuple,
+                                    ValidateEvent &&validate,
+                                    std::size_t const num_values)
```

Comment:
I have slight preference for this being combined with the other `run_basic_policy_test` overload, or call one of them `_impl`; they don't do the same thing, so they should not be overloads.

Reply:
These aren't overloads -- `run_basic_policy_tests` (plural) launches tests for each policy configurations by calling `run_basic_policy_test` (singular).

But I do like your idea of doing a function-scope try block in `basic_policy_test` instead of having the `run_basic_policy_test` trampoline, so I'll do that.